### ISSUE TITLE
Reinstall failed... my world has ended!

### ISSUE THREAD

**Issue by JangoBritt** [nfarina/homebridge#832]
So I was having some troubles with my homebridge, so I thought I'd do a full reinstall on my OSX Sierra machine.

I uninstalled all my modules
I uninstalled homebridge
I uninstalled node.js

re-downloaded node.js 4.6.0, installed successfully

When I go to install homebridge, I get the below gyp errors

MacMini:~ Adam Davidson$ sudo npm install -g homebridge
Password:
\
> curve25519-n@1.1.0 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/curve25519-n
> node-gyp rebuild

make: *** No rule to make target `../../../../../../../../../../Users/Adam', needed by `Makefile'.  Stop.
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 16.0.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/curve25519-n
gyp ERR! node -v v4.6.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 

> ed25519@0.0.4 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/ed25519
> node-gyp rebuild

make: *** No rule to make target `../../../../../../../../../../Users/Adam', needed by `Makefile'.  Stop.
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 16.0.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/ed25519
gyp ERR! node -v v4.6.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 
 
> mdns@2.3.3 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/mdns
> node-gyp rebuild

make: *** No rule to make target `../../../../../../../../../../Users/Adam', needed by `Makefile'.  Stop.
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 16.0.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/mdns
gyp ERR! node -v v4.6.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 

> bignum@0.12.5 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/srp/node_modules/bignum
> node-pre-gyp install --fallback-to-build

[bignum] Success: "/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/srp/node_modules/bignum/binding/bignum.node" is installed via remote
npm ERR! Darwin 16.0.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "homebridge"
npm ERR! node v4.6.0
npm ERR! npm  v2.15.9
npm ERR! code ELIFECYCLE

npm ERR! curve25519-n@1.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the curve25519-n@1.1.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the curve25519-n package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs curve25519-n
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls curve25519-n
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/Adam Davidson/npm-debug.log

**Comment by JangoBritt** [comment]
Ominously, my users path is "/Users/Adam Davidson" though the above log mentions only "/Users/Adam"... could there be an issue with the path name? If so, how did I get around it the first time I installed? lol

**Comment by KhaosT** [comment]
I'm not sure how you get through the first time but it's a known issue with npm that having a space in your user folder would prevent the install to finish... One way to fix this is to create a softlink for "/Users/Adam" and point that to your real user folder "/Users/Adam Davidson"...

**Comment by JangoBritt** [comment]
well... almost a eureka moment, that. I added the symlink, but now I get the same error, but with the second part of the string:

make: *** No rule to make target `Davidson/.node-gyp/4.6.0/include/node/common.gypi', needed by `Makefile'.  Stop.


Well Feck.

**Comment by KhaosT** [comment]
Try [this](https://github.com/nodejs/node-gyp/issues/65#issuecomment-179159383)?

**Comment by JangoBritt** [comment]
just tried that. Even though I've changed my $HOME to /AdamDavidson, it's still giving me the above message: make: *** No rule to make target Davidson/.node-gyp/4.6.0/include/node/common.gypi', needed byMakefile'. Stop

**Comment by KhaosT** [comment]
I'm not really sure then... maybe there are old files being cached somewhere for build information.

**Comment by JangoBritt** [comment]
well crap

I tried to remove absolutely everything that it installed to do a reinstall, but I guess I've missed something somewhere

**Comment by JangoBritt** [comment]
Any way to uninstall .node-gyp, and reinstall it somewhere else? looks like the problem is with the existing installation of node-gyp, and not when I try to install homebridge

**Comment by KhaosT** [comment]
this? https://github.com/nodejs/node-gyp/wiki/Updating-npm's-bundled-node-gyp

**Comment by JangoBritt** [comment]
nope. that didn't work.

I'm getting into dangerous territory now, hacking away at stuff I don't really know enough about. :-(

**Comment by JangoBritt** [comment]
ok... progress.

I scoured my drive, and found the node-gyp folder I wasn't deleting.

I've reinstalled node.js, and successfully (I think) installed homebridge.

I did get this error a couple of times:

gyp WARN EACCES user "root" does not have permission to access the dev dir "/Users/AdamDavidson/.node-gyp/4.6.0"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/mdns/.node-gyp"

Do I need to re-install using --unsafe-perm, or will it work without?

**Comment by KhaosT** [comment]
I think you need ``--unsafe-perm``

**Comment by JangoBritt** [comment]
looks like the only way I was able to reinstall it was without `--unsafe-perm`. Any other way and it was still recognising the literal path of the .node-gyp folder.

I've manage to successfully reinstall all my old modules (bar one) and it's working fine now, without the --unsafe-perm.

What functionality could I be losing doing it this way? 

I still have a problem with the HomeBridgeControllerLink, which is a shame, but can operate without it until I've raised the issue over on @KraigM's repo
